wire-ios-request-strategy/Sources/Payloads/Processing/ConversationEventPayloadProcessor.swift

```diff
@@ -484,54 +496,73 @@ final class ConversationEventPayloadProcessor {
+            let conversationType = payload.type.map(BackendConversationType.clientConversationType)
         else {
-            let conversation = ZMConversation.fetch(with: conversationID, domain: payload.qualifiedID?.domain, in: context)
-            conversation?.conversationType = self.conversationType(for: conversation, from: conversationType)
-            conversation?.needsToBeUpdatedFromBackend = false
-            return conversation
+            Logging.eventProcessing.error("Missing conversation or type in 1:1 conversation payload, aborting...")
+            return nil
         }
 
-        let otherUser = ZMUser.fetchOrCreate(with: otherUserID, domain: otherMember.qualifiedID?.domain, in: context)
-
-        var conversation: ZMConversation
-        if let existingConversation = otherUser.connection?.conversation {
-            existingConversation.mergeWithExistingConversation(withRemoteID: conversationID)
-            conversation = existingConversation
-        } else {
-            conversation = ZMConversation.fetchOrCreate(with: conversationID, domain: payload.qualifiedID?.domain, in: context)
-            otherUser.connection?.conversation = conversation
-        }
+        let conversation = ZMConversation.fetchOrCreate(
+            with: conversationID,
+            domain: payload.qualifiedID?.domain,
+            in: context
+        )
 
-        conversation.remoteIdentifier = conversationID
-        conversation.domain = BackendInfo.isFederationEnabled ? payload.qualifiedID?.domain : nil
         conversation.conversationType = self.conversationType(for: conversation, from: conversationType)
-
+        updateAttributes(from: payload, for: conversation, context: context)
+        updateMessageProtocol(from: payload, for: conversation)
         updateMetadata(from: payload, for: conversation, context: context)
         updateMembers(from: payload, for: conversation, context: context)
         updateConversationTimestamps(for: conversation, serverTimestamp: serverTimestamp)
         updateConversationStatus(from: payload, for: conversation)
-
         conversation.needsToBeUpdatedFromBackend = false
-        conversation.isPendingMetadataRefresh = otherUser.isPendingMetadataRefresh
+
+        if let otherUser = conversation.localParticipantsExcludingSelf.first {
+            if
+                let connection = otherUser.connection,
+                let existingConversation = connection.conversation,
+                existingConversation.messageProtocol == .proteus,
+                conversation.messageProtocol == .mls
+            {
+                // Invalidate old proteus conversation, make mls conversation active.
+                existingConversation.conversationType = .invalid
+                connection.conversation = conversation
+            }
+
+            conversation.isPendingMetadataRefresh = otherUser.isPendingMetadataRefresh
+        }
```

Comment:
> Shouldn't we also check what common supported protocols are before we switch to use the MLS 1-1 conversation?

@typfel this is assuming that if we become aware of the mls one on one then it's because we've been added to the group and the supported protocol check was already done. But indeed I see that it may be a flaky assumption to make. I guess this goes back to our previous discussion about leaving the payload processor for only updating the local objects from remote objects and finding a better place to move the logic of switching the mls one on one.

I'll see if I can move it to where we receive the welcome message.

Reply:
Refactored